### PR TITLE
Stock transfers mgmt

### PR DESCRIPTION
**Pull request by allisonlarson**
Beginning of adding more administrable functionality for stock transfers; currently builds out the initial modeling for stock transfers and added transfer items, and provides an index for viewing based on stock locations.

* Stock Transfers have shipped_at, tracking number, and closed_at fields
* Stock Transfers have transfer items, which can be indicated as being received_at
* Transfer Items belongs to variant, stock transfer, and stock location
* Stock Locations have many transfer items & specs

* Added searchable index view for stock transfers
* Stock Transfer controller inherits from resource controller
* Added translations
* Added and updated specs

### CODE REVIEW

File: backend/app/controllers/spree/admin/stock_transfers_controller.rb [hunk 1]
```diff
@@ -35,11 +18,20 @@ def create
         redirect_to admin_stock_transfer_path(stock_transfer)
       end
 
-      private
-      def load_stock_locations
-        @stock_locations = Spree::StockLocation.active.order_default
+      def set_stock_locations
```
Comment (richardnuno) [inline backend/app/controllers/spree/admin/stock_transfers_controller.rb]: Convention for these types of before_filters seems to be `load_*`, not really partial to either one but it'd be good to keep with the convention

File: backend/app/controllers/spree/admin/stock_transfers_controller.rb [hunk 2]
```diff
@@ -35,11 +18,20 @@ def create
...
+       @stock_locations = Spree::StockLocation.accessible_by(current_ability, :index)
```
Comment (richardnuno) [inline backend/app/controllers/spree/admin/stock_transfers_controller.rb]: Indentation looks a little off on this

File: backend/app/controllers/spree/admin/stock_transfers_controller.rb [hunk 3]
```diff
@@ -35,11 +18,20 @@ def create
...
+      end
+
+      def collection
```
Comment (richardnuno) [inline backend/app/controllers/spree/admin/stock_transfers_controller.rb]: Any reason why `collection` and `set_stock_locations` is public?

File: backend/app/views/spree/admin/stock_transfers/index.html.erb [hunk 4]
```diff
@@ -10,76 +8,97 @@
...
 
-      <div class="omega four columns">
+      <div class="field-block alpha four columns">
         <div class="field">
-          <%= f.label :destination_location, Spree.t(:destination) %>
-          <%= f.select :destination_location_id_eq,
-               options_from_collection_for_select(@stock_locations, :id, :name, @q.destination_location_id_eq),
-               { include_blank: true }, class: 'select2 fullwidth' %>
+          <%= f.label nil, Spree.t(:status) %>
+          <%= f.select :closed_at_null, options_for_select([[Spree.t(:open), 1],[Spree.t(:closed), 0]], params[:q][:closed_at_null]), {include_blank: true}, {class: 'select2 fullwidth'} %>
```
Comment (richardnuno) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: Is the `_null` for ransack?

Comment (richardnuno) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: Ah forget it, got it now...it's a boolean value

File: backend/config/routes.rb [hunk 5]
```diff
@@ -149,7 +149,7 @@
 
     resources :shipping_methods
     resources :shipping_categories
-    resources :stock_transfers, :only => [:index, :show, :new, :create]
+    resources :stock_transfers, :only => [:index, :show, :new, :create, :edit]
```
Comment (richardnuno) [inline backend/config/routes.rb]: This should probably be part of another PR

File: backend/app/views/spree/admin/stock_transfers/index.html.erb (hunk 4 above)
Comment (allisonlarson) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: yeah, well `closed_at` itself isn't a boolean value, but we want to be able to search by if a stock transfer is open or closed (or if there is a value in closed_at or not)

File: backend/app/controllers/spree/admin/stock_transfers_controller.rb (hunk 2 above)
Comment (allisonlarson) [inline backend/app/controllers/spree/admin/stock_transfers_controller.rb]: Yep, thanks

File: backend/app/controllers/spree/admin/stock_transfers_controller.rb (hunk 1 above)
Comment (allisonlarson) [inline backend/app/controllers/spree/admin/stock_transfers_controller.rb]: Yeah, it makes sense to keep load conventions.

File: core/app/models/spree/stock_transfer.rb [hunk 6]
```diff
@@ -1,12 +1,20 @@
...
 
     belongs_to :source_location, :class_name => 'StockLocation'
     belongs_to :destination_location, :class_name => 'StockLocation'
 
+    scope :by_location, ->(location_id){ where('source_location_id=? OR destination_location_id=?', location_id, location_id)}
+
     make_permalink field: :number, prefix: 'T'
 
+
+    def status
```
Comment (richardnuno) [inline core/app/models/spree/stock_transfer.rb]: This seems like more of a view concern. Perhaps a `closed?` method on the model would make more sense and use that in the view to determine the status to display.

File: core/lib/spree/testing_support/factories/stock_transfer_factory.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+FactoryGirl.define do
+  factory :stock_transfer, class: Spree::StockTransfer do
+    source_location Spree::StockLocation.first
```
Comment (richardnuno) [inline core/lib/spree/testing_support/factories/stock_transfer_factory.rb]: Depending on whether a stock location has already been created or not, this might assign nil to the source_location. Also, if there's only one stock location the source and destination could be the same. Creating two new stock_locations might be safer in this case.

File: backend/app/views/spree/admin/stock_transfers/index.html.erb [hunk 8]
```diff
@@ -10,76 +8,97 @@
...
         </div>
       </div>
 
       <div class="clearfix"></div>
 
       <div class="actions filter-actions">
-        <div data-hook="admin_stock_transfers_index_search_buttons">
+        <div data-hook="admin_orders_index_search_buttons">
```
Comment (richardnuno) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: Any special reason for this change? Previous value seemed correct

File: backend/app/views/spree/admin/stock_transfers/index.html.erb [hunk 9]
```diff
@@ -10,76 +8,97 @@
...
-<div data-hook="admin_orders_index_search">
-  <fieldset>
-    <legend align="center"><%= Spree.t(:search) %></legend>
-    <%= search_form_for @q, :url => admin_stock_transfers_path do |f| %>
+<% content_for :table_filter_title do %>
+  <%= Spree.t(:search) %>
+<% end %>
 
+<% content_for :table_filter do %>
+  <div data-hook="admin_orders_index_search">
```
Comment (richardnuno) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: Probably want to rename this data-hook

File: backend/app/views/spree/admin/stock_transfers/index.html.erb [hunk 10]
```diff
@@ -10,76 +8,97 @@
...
 <% if @stock_transfers.any? %>
-  <table class='index' id='listing_stock_transfers' data-hook>
-    <colgroup>
-      <col style='width: 25%' />
-      <col style='width: 20%' />
-      <col style='width: 20%' />
-      <col style='width: 25%' />
-      <col style='width: 10%' />
-    </colgroup>
+  <table class="index" id="listing_transfer_orders">
```
Comment (richardnuno) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: We probably want to keep with the stock_transfers name instead of transfer_orders

Comment (allisonlarson) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: Whoops, went through some name changes and didn't catch this one.

File: core/app/models/spree/stock_transfer.rb (hunk 6 above)
Comment (allisonlarson) [inline core/app/models/spree/stock_transfer.rb]: I agree, I think adding the status to a helper would be the way to go here.

File: backend/app/views/spree/admin/stock_transfers/index.html.erb [hunk 11]
```diff
@@ -87,9 +106,9 @@
   </table>
 <% else %>
   <div class="alpha twelve columns no-objects-found">
```
Comment (richardnuno) [inline backend/app/views/spree/admin/stock_transfers/index.html.erb]: Just noticed while trying this out locally, when there are no results this should be sixteen columns instead of twelve

File: backend/app/helpers/spree/admin/stock_transfers_helper.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+      def handle_stock_transfer(stock_transfer)
+        if can?(:show, stock_transfer)
+          link_to stock_transfer.number, admin_stock_transfer_path(stock_transfer)
+        else
+          stock_transfer.number
+        end
+      end
+
+      def status(stock_transfer)
+        stock_transfer.closed? ? "Closed" : "Open"
```
Comment (gmacdougall) [inline backend/app/helpers/spree/admin/stock_transfers_helper.rb]: These values should be translated.

File: backend/app/controllers/spree/admin/stock_transfers_controller.rb [hunk 13]
```diff
@@ -36,8 +19,17 @@ def create
       end
 
       private
+
       def load_stock_locations
-        @stock_locations = Spree::StockLocation.active.order_default
+        @stock_locations = Spree::StockLocation.accessible_by(current_ability, :index)
+      end
+
+      def collection
```
Comment (gmacdougall) [inline backend/app/controllers/spree/admin/stock_transfers_controller.rb]: I would prefer that this maintain the visibility inherited from the parent which is `protected`. It's a bit confusing to see a private method not used in a file.

File: core/app/models/spree/stock_location.rb [hunk 14]
```diff
@@ -17,6 +17,10 @@ class StockLocation < ActiveRecord::Base
     after_create :create_stock_items, :if => "self.propagate_all_variants?"
     after_save :ensure_one_default
 
+    def stock_transfers
+      Spree::StockTransfer.by_location(self.id)
+    end
```
Comment (gmacdougall) [inline core/app/models/spree/stock_location.rb]: This should probably use a rails `has_many` relationship?

Comment (gmacdougall) [inline core/app/models/spree/stock_location.rb]: On second thought, the `by_location` dealing with source or destination location makes that a bit weird. At a minimum we should have some better relationships between `StockLocation` and `StockTransfer`.

Comment (allisonlarson) [inline core/app/models/spree/stock_location.rb]: Agreed, very open to rethinking the relationship between StockLocation and StockTransfer. A `has_many` relationship isn't possible with multiple foreign keys from what I could find, so I'd be curious to other solutions where StockTransfer still has a source and destination location but that would follow more conventional associations.

File: backend/app/helpers/spree/admin/stock_transfers_helper.rb (hunk 12 above)
Comment (allisonlarson) [inline backend/app/helpers/spree/admin/stock_transfers_helper.rb]: Thanks for catching

File: backend/app/controllers/spree/admin/stock_transfers_controller.rb (hunk 13 above)
Comment (allisonlarson) [inline backend/app/controllers/spree/admin/stock_transfers_controller.rb]: Agreed, that makes more sense.

File: core/app/models/spree/stock_location.rb (hunk 14 above)
Comment (allisonlarson) [inline core/app/models/spree/stock_location.rb]: Actually deciding to keep with source/destination locations on stock transfer and keep the relationship the same.

File: core/app/models/spree/stock_transfer.rb [hunk 15]
```diff
@@ -1,12 +1,17 @@
...
 
-    belongs_to :source_location, :class_name => 'StockLocation'
-    belongs_to :destination_location, :class_name => 'StockLocation'
+    belongs_to :source_location, :class_name => 'Spree::StockLocation'
+    belongs_to :destination_location, :class_name => 'Spree::StockLocation'
 
     make_permalink field: :number, prefix: 'T'
 
+    def closed?
+      !closed_at.nil?
```
Comment (richardnuno) [inline core/app/models/spree/stock_transfer.rb]: You can avoid the ! by using `closed_at.present?`

File: core/db/migrate/20150407173305_add_fields_to_stock_transfer.rb [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class AddFieldsToStockTransfer < ActiveRecord::Migration
+  def change
+    add_column :spree_stock_transfers, :shipped_at, :datetime
+    add_column :spree_stock_transfers, :closed_at, :datetime
+    add_column :spree_stock_transfers, :tracking_number, :string
```
Comment (richardnuno) [inline core/db/migrate/20150407173305_add_fields_to_stock_transfer.rb]: We're probably going to need another table for the tracking number since we need to support multiple tracking number per transfer.

Comment (richardnuno) [inline core/db/migrate/20150407173305_add_fields_to_stock_transfer.rb]: Might be good to add a `created_by` and a `closed_by` field.

File: core/db/migrate/20150407173531_create_transfer_items.rb [hunk 17]
```diff
@@ -0,0 +1,13 @@
+class CreateTransferItems < ActiveRecord::Migration
+  def change
+    create_table :spree_transfer_items do |t|
+      t.integer :variant_id
+      t.integer :stock_transfer_id
+      t.datetime :received_at
```
Comment (richardnuno) [inline core/db/migrate/20150407173531_create_transfer_items.rb]: And a `received_by` here although not sure if it's that valuable here

File: core/app/models/spree/stock_transfer.rb (hunk 15 above)
Comment (athal7) [inline core/app/models/spree/stock_transfer.rb]: or !!closed_at

File: core/db/migrate/20150407173305_add_fields_to_stock_transfer.rb (hunk 16 above)
Comment (allisonlarson) [inline core/db/migrate/20150407173305_add_fields_to_stock_transfer.rb]: Are you thinking `created_by` and `closed_by` would be strings, or references to users? I can't think of any case where we would want them to not be users, but just want to make sure.

File: core/app/models/spree/stock_transfer.rb (hunk 15 above)
Comment (allisonlarson) [inline core/app/models/spree/stock_transfer.rb]: `closed_at.present?` was on the tip of my tongue.

File: core/db/migrate/20150407173531_create_transfer_items.rb (hunk 17 above)
Comment (allisonlarson) [inline core/db/migrate/20150407173531_create_transfer_items.rb]: Would there be cases where different people would receive than who would close?

File: core/db/migrate/20150407173305_add_fields_to_stock_transfer.rb (hunk 16 above)
Comment (allisonlarson) [inline core/db/migrate/20150407173305_add_fields_to_stock_transfer.rb]: After talking through, going to keep tracking number as a string instead of another table.

File: core/app/models/spree/stock_transfer.rb [hunk 18]
```diff
@@ -1,16 +1,31 @@
...
+    def closed?
+      closed_at.present?
+    end
+
     def to_param
       number
     end
 
+    def received_item_count
+      transfer_items.inject(0) { |sum, item| sum + item.received_quantity }
```
Comment (gmacdougall) [inline core/app/models/spree/stock_transfer.rb]: You don't need to inject 0. 0 is the default argument.

You also may want to get this information through a SQL query instead of loading up a bunch of ruby objects.

```
transfer_items.sum(:received_quantity)
```

Comment (allisonlarson) [inline core/app/models/spree/stock_transfer.rb]: in this instance, the default argument is actually a TransferItem. :confused: I tried it initially without injecting 0.

But yes, a sql query sounds much better.

**Comment by allisonlarson** [comment]
@gmacdougall I changed the stock transfer's expected and received items to use sql queries, is there anything else that sticks out to you?

File: core/app/models/spree/transfer_item.rb [hunk 19]
```diff
@@ -0,0 +1,9 @@
+module Spree
+  class TransferItem < ActiveRecord::Base
```
Comment (gmacdougall) [inline core/app/models/spree/transfer_item.rb]: There should be some validations on this. A `TransferItem` with a bunch of nulls probably isn't a valid representation of a transfer item.

File: core/db/migrate/20150407173531_create_transfer_items.rb [hunk 20]
```diff
@@ -0,0 +1,14 @@
+class CreateTransferItems < ActiveRecord::Migration
+  def change
+    create_table :spree_transfer_items do |t|
+      t.integer :variant_id
+      t.integer :stock_transfer_id
+      t.integer :expected_quantity, default: 0
+      t.integer :received_quantity, default: 0
```
Comment (gmacdougall) [inline core/db/migrate/20150407173531_create_transfer_items.rb]: Columns which are required should be flagged `null: false` here.

**Comment by gmacdougall** [comment]
Everything looks pretty good aside from some likely missing validations. Sorry I missed those on my first couple passes.

File: core/app/models/spree/transfer_item.rb (hunk 19 above)
Comment (allisonlarson) [inline core/app/models/spree/transfer_item.rb]: Yep, should have caught these validation ones. Thanks!

**Comment by allisonlarson** [comment]
@gmacdougall Added in those validations

File: core/db/migrate/20150407173531_create_transfer_items.rb [hunk 21]
```diff
@@ -0,0 +1,14 @@
+class CreateTransferItems < ActiveRecord::Migration
+  def change
+    create_table :spree_transfer_items do |t|
+      t.integer :variant_id
+      t.integer :stock_transfer_id
```
Comment (gmacdougall) [inline core/db/migrate/20150407173531_create_transfer_items.rb]: `:variant_id` and `:stock_transfer_id` can also be set to `null: false` since they should always have value.

Maintaining the validations at the DB level ensures that our data is always consistent (and can't be bypassed), the validations at the rails level tend to provide nicer errors.

Comment (allisonlarson) [inline core/db/migrate/20150407173531_create_transfer_items.rb]: Interesting, I didn't think about that. Thanks

**Comment by allisonlarson** [comment]
@gmacdougall hopefully last ping?

**Comment by gmacdougall** [comment]
:+1:

**Comment by richardnuno** [comment]
:+1:
